v3_async_wip/v3_async_wip/iothub_mqtt_client.py

```diff
@@ -0,0 +1,627 @@
+
+import asyncio
+import json
+import logging
+import urllib.parse
+from typing import Optional, AsyncGenerator
+from .custom_typing import TwinPatch, Twin
+from .models import Message, MethodResponse, MethodRequest
+from . import config, constant
+from . import request_response as rr
+from . import mqtt_client as mqtt
+from azure.iot.device import user_agent  # type: ignore
+from azure.iot.device.iothub.pipeline import mqtt_topic_iothub as mqtt_topic  # type: ignore
+from azure.iot.device.common.auth import sastoken as st  # type: ignore
+from azure.iot.device.common import alarm  # type: ignore
+
+# TODO: add typings to reused V2 code
+# TODO: update docstrings with correct class paths once repo structured better
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RECONNECT_INTERVAL = 10
+DEFAULT_TOKEN_UPDATE_MARGIN = 120
+
+# TODO: add exceptions to docstring
+# TODO: background exceptions how
+# TODO: non-background exceptions
+
+
+class IoTHubError(Exception):
+    """Represents a failure reported by IoTHub"""
+
+    pass
+
+
+class IoTHubMQTTClient:
+    def __init__(
+        self,
+        client_config: config.IoTHubClientConfig,
+    ) -> None:
+        """
+        Instantiate the client
+
+        :param client_config: The config object for the client
+        :type client_config: :class:`IoTHubClientConfig`
+        """
+        # Basic Information
+        self._device_id = client_config.device_id
+        self._module_id = client_config.module_id
+
+        # Sastoken Auth
+        self._sastoken: Optional[st.SasToken]
+        self._sastoken_update_alarm: Optional[alarm.Alarm]
+        if client_config.sastoken is not None:
+            self._sastoken = client_config.sastoken
+            self._sastoken_update_alarm = self._create_token_update_alarm()
+        else:
+            self._sastoken = None
+            self._sastoken_update_alarm = None
+
+        # MQTT Configuration
+        self._mqtt_client = _create_mqtt_client(client_config)
+
+        # Create incoming IoTHub data generators
+        self.incoming_c2d_messages: AsyncGenerator[Message, None] = _create_c2d_message_generator(
+            self._device_id, self._mqtt_client
+        )
+        self.incoming_input_messages: Optional[
+            AsyncGenerator[Message, None]
+        ] = _create_input_message_generator(self._device_id, self._module_id, self._mqtt_client)
+        self.incoming_method_requests: AsyncGenerator[
+            MethodRequest, None
+        ] = _create_method_request_generator(self._mqtt_client)
+        self.incoming_twin_patches: AsyncGenerator[TwinPatch, None] = _create_twin_patch_generator(
+            self._mqtt_client
+        )
+
+        # Internal request/response infrastructure
+        self._request_ledger = rr.RequestLedger()
+        self._twin_responses_enabled = False
+        self._twin_response_listener = asyncio.create_task(self._process_twin_responses())
```

Comment:
>_twin_response_listener [](http://example.com/codeflow?start=13&length=23)

What happens if this task raises an exception?

Reply:
It will leverage the not-yet-existing infrastructure for background error reporting (a couple PRs from now). Will add a TODO comment to reflect this so it doesn't get forgotten.